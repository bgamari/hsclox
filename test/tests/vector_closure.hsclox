fun vector(a1, a2) {
    var x = a1;
    var y = a2;
    fun get(field) {
        switch (field) {
            case "x": return x;
            case "y": return y;
            default: return nil;
        }
    }
    return get;
}

fun addVectors(va, vb) {
    return vector(va("x") + vb("x"), va("y") + vb("y"));
}

var v1 = vector(3, 6);
var v2 = vector(7, 4);

print v1("x");
print v1("y");
print v2("x");
print v2("y");

var vsum = addVectors(v1, v2);

print vsum("x");
print vsum("y");
