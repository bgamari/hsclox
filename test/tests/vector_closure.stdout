== get ==
0000    5 OP_GET_LOCAL        1
0002    | OP_STACK_DUP_1
0003    6 OP_CONSTANT         0 'x
0005    | OP_EQUAL
0006    | OP_JUMP_IF_FALSE    6 -> 17
0009    | OP_POPN             2 'false
0011    | OP_GET_UPVALUE      0
0013    | OP_RETURN
0014    | OP_JUMP            14 -> 37
0017    | OP_POP
0018    | OP_STACK_DUP_1
0019    7 OP_CONSTANT         1 'y
0021    | OP_EQUAL
0022    | OP_JUMP_IF_FALSE   22 -> 33
0025    | OP_POPN             2 'false
0027    | OP_GET_UPVALUE      1
0029    | OP_RETURN
0030    | OP_JUMP            30 -> 37
0033    | OP_POPN             2 'false
0035    8 OP_NIL
0036    | OP_RETURN
0037   10 OP_NIL
0038    | OP_RETURN
== vector ==
0000    2 OP_GET_LOCAL        1
0002    3 OP_GET_LOCAL        2
0004   10 OP_CLOSURE          0 <fn get>
0006     |                    local 3
0008     |                    local 4
0010   11 OP_GET_LOCAL        5
0012    | OP_RETURN
0013   12 OP_NIL
0014    | OP_RETURN
== addVectors ==
0000   15 OP_GET_GLOBAL       0 'vector
0002    | OP_GET_LOCAL        1
0004    | OP_CONSTANT         1 'x
0006    | OP_CALL             1
0008    | OP_GET_LOCAL        2
0010    | OP_CONSTANT         2 'x
0012    | OP_CALL             1
0014    | OP_ADD
0015    | OP_GET_LOCAL        1
0017    | OP_CONSTANT         3 'y
0019    | OP_CALL             1
0021    | OP_GET_LOCAL        2
0023    | OP_CONSTANT         4 'y
0025    | OP_CALL             1
0027    | OP_ADD
0028    | OP_CALL             2
0030    | OP_RETURN
0031   16 OP_NIL
0032    | OP_RETURN
== <script> ==
0000   12 OP_CLOSURE          1 <fn vector>
0002    | OP_DEFINE_GLOBAL    0 'vector
0004   16 OP_CLOSURE          3 <fn addVectors>
0006    | OP_DEFINE_GLOBAL    2 'addVectors
0008   18 OP_GET_GLOBAL       0 'vector
0010    | OP_CONSTANT         5 '3
0012    | OP_CONSTANT         6 '6
0014    | OP_CALL             2
0016    | OP_DEFINE_GLOBAL    4 'v1
0018   19 OP_GET_GLOBAL       0 'vector
0020    | OP_CONSTANT         8 '7
0022    | OP_CONSTANT         9 '4
0024    | OP_CALL             2
0026    | OP_DEFINE_GLOBAL    7 'v2
0028   21 OP_GET_GLOBAL       4 'v1
0030    | OP_CONSTANT        10 'x
0032    | OP_CALL             1
0034    | OP_PRINT
0035   22 OP_GET_GLOBAL       4 'v1
0037    | OP_CONSTANT        11 'y
0039    | OP_CALL             1
0041    | OP_PRINT
0042   23 OP_GET_GLOBAL       7 'v2
0044    | OP_CONSTANT        12 'x
0046    | OP_CALL             1
0048    | OP_PRINT
0049   24 OP_GET_GLOBAL       7 'v2
0051    | OP_CONSTANT        13 'y
0053    | OP_CALL             1
0055    | OP_PRINT
0056   26 OP_GET_GLOBAL       2 'addVectors
0058    | OP_GET_GLOBAL       4 'v1
0060    | OP_GET_GLOBAL       7 'v2
0062    | OP_CALL             2
0064    | OP_DEFINE_GLOBAL   14 'vsum
0066   28 OP_GET_GLOBAL      14 'vsum
0068    | OP_CONSTANT        15 'x
0070    | OP_CALL             1
0072    | OP_PRINT
0073   29 OP_GET_GLOBAL      14 'vsum
0075    | OP_CONSTANT        16 'y
0077    | OP_CALL             1
0079    | OP_PRINT
0080   30 OP_NIL
0081    | OP_RETURN
3
6
7
4
10
10
